### PR TITLE
backend-plugin-api: allow modules to register extension points

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

After a lot of debating we're leaning towards this solution for how to improve module options.

TL;DR of the problem: We want plugins and modules to be installable via declarative integration, which means no code. This goes against our current design with plugins and modules being exported as factories that accept options for further customization. The options are alright for static values, if a bit awkward due to the duplication, but they work very poorly for values that can only be expressed in code, e.g. callbacks. Options are also kind of bad already, because they don't have access to service dependencies, which is something we want to fix too.

This proposed solution is to allow for modules to declare and export extension points, which can then be consumed by modules for the same plugin. These second level modules would typically always be internal customizations, so we don't intend for module packages to need their own `-node` library packages as an indirection for module extension points.

This PR also ports the incremental ingestion provider to this new pattern to show some usage. Here's what that looks like:

```ts
import {
  catalogModuleIncrementalIngestionEntityProvider,
  incrementalIngestionProvidersExtensionPoint,
} from '@backstage/plugin-catalog-backend-module-incremental-ingestion';

backend.add(catalogModuleIncrementalIngestionEntityProvider());
backend.add(
  createBackendModule({
    pluginId: 'catalog',
    moduleId: 'incrementalTest',
    register(env) {
      env.registerInit({
        deps: { extension: incrementalIngestionProvidersExtensionPoint },
        async init({ extension }) {
          extension.addProvider({
            provider: myCustomProvider,
            options: myCustomProviderOptions,
          });
        },
      });
    },
  }),
);
```

Note that we currently require `createBackendModule(...)()`, but that's being fixed in a separate PR.

Figured we should leave documentation updates for a separate PR, since we also want to address removal of options, service factories as features, etc.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-plugin-api/src/wiring/factories.ts [hunk 1]
```diff
@@ -194,10 +195,20 @@ export function createBackendModule<TOptions extends [options?: object] = []>(
...
+        const extensionPoints: InternalBackendPluginRegistration['extensionPoints'] =
+          [];
         let init: InternalBackendModuleRegistration['init'] | undefined =
           undefined;
 
         c.register({
+          registerExtensionPoint(ext, impl) {
+            if (init) {
+              throw new Error(
+                'registerExtensionPoint called after registerInit',
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-plugin-api/src/wiring/factories.ts:207]: It's always been this amount of needed :p

I think we should keep it for now tbh, in case we want it to matter in the future.
